smart-wallets/src/token/AssetToken.sol

```diff
@@ -145,11 +150,19 @@ contract AssetToken is WalletUtils, YieldDistributionToken, IAssetToken {
             if (getBalanceAvailable(from) < value) {
                 revert InsufficientBalance(from);
             }
+
+            if (balanceOf(from) == value) {
+                // Will have zero balance after transfer
+                emit HolderStatusChanged(from, false);
+            }
         }
 
-        if (!$.hasHeld[to]) {
-            $.holders.push(to);
-            $.hasHeld[to] = true;
+        if (to != address(0)) {
+            // Check if to address will become a new holder
+            if (balanceOf(to) == 0) {
+                // Currently has zero balance
+                emit HolderStatusChanged(to, true);
```

Comment:
I think we should keep `holders` and `hasHeld` and remove the `HolderStatusChanged`

Reply:
we can get those info from events? 
holders will contain every address that has ever held tokens, even if they no longer do and arrays consume unbounded storage, gas costs increase linearly with array size